Add support for installing `openhab` & `openhab_rules_tools` from npm

florian-h05 opened the pull request:
Fixes #1651.

This PR adds a menu option to install the openHAB JavaScript library and `openhab_rules_tools` from @rkoshak.

florian-h05 commented:
@mstormi 
This is nearly ready for merging.

At every start of the `openhabian-config` tool, it checks the following for both `openhab` & `openhab_rules_tools`:
1. Does the package folder exist? If not, exit.
2. Run `npm outdated` and check for the package name in the output. If the package is mentioned in the output, an update is available.
3. Ask the user for updating.

In the `openhabian-config` tool, you can install both packages under _40 | openHAB Related_.

I only need your help where to put the command for auto-installing `openhab_rules_tools`.

mstormi commented on functions/menu.bash:
```diff
@@ -199,6 +199,8 @@ show_main_menu() {
     "42 | Upgrade to openHAB 3"   "Upgrade OS environment to openHAB 3 release" \
     "   | Downgrade to openHAB 2" "Downgrade OS environment from openHAB 3 back to openHAB 2 (DANGEROUS)" \
     "43 | Remote Console"         "Bind the openHAB SSH console to all external interfaces" \
+    "   | openhab-js"             "JS Scripting: Upgrade to the latest version of the openHAB JavaScript library" \
```
use own menu number. Unnumbered lines are options to the previous numbered line

mstormi commented on openhabian-setup.sh:
```diff
@@ -132,6 +132,8 @@ else
   load_create_config
   openhabian_console_check
   openhabian_update_check
+  jsscripting_npm_check "openhab"
```
wrong branch use the other one ($UNATTENDED is set for  unattended (=initial) installations so this is where your auto-install must go)

rkoshak commented on functions/menu.bash:
```diff
@@ -199,6 +199,8 @@ show_main_menu() {
     "   | Downgrade to openHAB 2" "Downgrade OS environment from openHAB 3 back to openHAB 2 (DANGEROUS)" \
     "43 | Remote Console"         "Bind the openHAB SSH console to all external interfaces" \
+    "50 | openhab-js"             "JS Scripting: Upgrade to the latest version of the openHAB JavaScript library" \
+    "51 | openhab_rules_tools"    "JS Scripting: Install the latest version of openhab_rules_tools" \
```
If we are installing by default, I don't think we need menu option 51 any longer

florian-h05 replied:
Makes sense, thanks.

rkoshak commented on functions/menu.bash:
```diff
@@ -216,6 +218,8 @@ show_main_menu() {
       *Downgrade\ to\ openHAB\ 2) migrate_installation "openHAB2" && openhabian_update "stable";;
       43\ *) openhab_shell_interfaces;;
+      50\ *) jsscripting_npm_install "openhab";;
+      51\ *) jsscripting_npm_install "openhab_rules_tools";;
```
Same comment, if it's being installed by default we probably don't need this menu option.

rkoshak commented on functions/nodejs-apps.bash:
```diff
@@ -342,3 +342,60 @@ zigbee2mqtt_setup() {
+  if [ "${2}" == "update" ];
+  then
+    echo -n "$(timestamp) [openHABian] Updating ${1} for JS Scripting... "
+    if cond_redirect sudo -u "${username:-openhabian}" npm update --prefix "/etc/openhab/automation/js" "${1}"; then echo "OK"; else echo "FAILED (npm update)"; return 1; fi
```
Do we have to worry about file permissions here? At some point openHAB will change the ownership of everything in `/etc/openhab/` to `openhab:openhab`. It might cause problems to update/install as user `openhabian` or it might cause problems for openHAB after the update is complete.

florian-h05 replied:
That‘s a good point, thanks!
I am nut sure if this can cause trouble (depends on the chmod number), but even if not, it is the better way to install as openhab:openhab.
Why should we install as `openhabian` when it anyway gets modified to `openhab:openhab` later?

rkoshak reviewed (commented):
Just a few quick comments

florian-h05 commented:
@rkoshak @mstormi 
Sorry for the chaos with the review requests.
I‘m somehow not able to request a re-review from both of you using the „request review“ button.

rkoshak commented:
Short of being able to test this myself all I can say is LGTM.

florian-h05 commented:
I tested all recent changes just a few minutes ago, everything works.

Should be ready for merging @mstormi.

mstormi commented on functions/menu.bash:
```diff
@@ -216,6 +217,7 @@ show_main_menu() {
       42\ *) migrate_installation "openHAB3" && openhabian_update "openHAB3";;
       *Downgrade\ to\ openHAB\ 2) migrate_installation "openHAB2" && openhabian_update "stable";;
       43\ *) openhab_shell_interfaces;;
+      50\ *) jsscripting_npm_install "openhab";;
```
order

mstormi commented:
please rebase to include a fix I've just to main (it should make the installation tests pass)

mstormi commented on functions/nodejs-apps.bash:
```diff
@@ -342,3 +342,60 @@ zigbee2mqtt_setup() {
+##    Available values for mode: "update", "install". Defaults to "install".
+##
+jsscripting_npm_install() {
+  if [ "${1}" == "" ]; then echo "FAILED. Provide packageName."; return 1; fi
```
Please check for $UNATTENDED and distinguish between interactive install (chosen via menu) and initial (unattended) install.
If interactive, provide feedback via whiptail info boxes.
If unattended, be verbose about what you're doing
(like echo "Installing foobar ...\c" ..... echo "OK" ..... )

See how installationb is done for other packages/function for examples what I mean by that.

Apply to uninstalling, too.

mstormi commented on functions/nodejs-apps.bash:
```diff
@@ -342,3 +342,60 @@ zigbee2mqtt_setup() {
+    if cond_redirect nodejs_setup; then echo "OK"; else echo "FAILED"; return 1; fi
+  fi
+
+  introText="Additions, improvements or fixes were added to ${1} (npm package) for JS Scripting. Would you like to update now and benefit from them?\\nThe update might include breaking changes, please head over to the JS Scripting docs or to https://www.npmjs.com/package/${1}.\\n\\nUpdating is recommended."
```
Mind the openHABian conventions (see CONTRIBUTING.md). Define consts like this as local at the beginning of the routine.

mstormi commented on functions/nodejs-apps.bash:
```diff
@@ -342,3 +342,60 @@ zigbee2mqtt_setup() {
+  introText="Additions, improvements or fixes were added to ${1} (npm package) for JS Scripting. Would you like to update now and benefit from them?\\nThe update might include breaking changes, please head over to the JS Scripting docs or to https://www.npmjs.com/package/${1}.\\n\\nUpdating is recommended."
+
+  echo -n "$(timestamp) [openHABian] Checking for updates of ${1} for JS Scripting... "
+  local outdatedReturn
```
local to top

mstormi commented on openhabian-setup.sh:
```diff
@@ -126,12 +126,15 @@ if [[ -n "$UNATTENDED" ]]; then
   permissions_corrections
   setup_mirror_SD "install"
   install_cleanup
+  jsscripting_npm_install "openhab_rules_tools"
```
move to a more adequate location (after frontail?)

mstormi reviewed (changes requested):
please also rebase

florian-h05 commented:
@mstormi 
I addressed your review and rebased, retested all current changes and everything works.
From my side this is ready for merging, feel free to re-review.

mstormi reviewed (changes requested):
almost there

florian-h05 commented:
@mstormi 
I updated the whiptail menu size:

![IMG_1715](https://user-images.githubusercontent.com/73423173/191603572-9d844222-af53-443b-b17e-8d4d60ff5d7a.png)

mstormi reviewed (approved):
LGTM thanks

mstormi commented:
There's a branch inside openhabian-setup.sh that's executed for unattended installs. Put it somewhere after openhab installation.Am 13.09.2022 15:24 schrieb Florian Hotze ***@***.***>:
@mstormi
This is nearly ready for merging.
At every start of the openhabian-config tool, it checks the following for both openhab & openhab_rules_tools:
Does the package folder exist? If not, exit.Run npm outdated and check for the package name in the output. If the package is mentioned in the output, an update is available.Ask the user for updating.
In the openhabian-config tool, you can install both packages under 40 | openHAB Related.
I only need your help where to put the command for auto-installing openhab_rules_tools.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>
